### PR TITLE
[CP Staging] Fix: No notification when closing account

### PR DESCRIPTION
**Pull request by tienifr**
### Details

No notification when account is closed. This PR fixes that.

### Fixed Issues

$ https://github.com/Expensify/App/issues/28748
PROPOSAL: https://github.com/Expensify/App/issues/28748#issuecomment-1746091428

### Tests

1. Login with any account
2. Go to Settings >> Security >> Close account
3. Enter your default contact method >> Close account >> Confirm
4. Verify you are logged out and a notification says "Account closed successfully"

- [x] Verify that no errors appear in the JS console

### Offline tests

NA

### QA Steps

1. Login with any account
2. Go to Settings >> Security >> Close account
3. Enter your default contact method >> Close account >> Confirm
4. Verify you are logged out and a notification says "Account closed successfully"

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://github.com/Expensify/App/assets/113963320/59e65f31-1aa8-4c6c-a570-4f856f48c464

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://github.com/Expensify/App/assets/113963320/c576ebfe-ece4-4a6f-b51b-31560e0255fd

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://github.com/Expensify/App/assets/113963320/d4f526bb-29aa-44fa-be8b-d9dbbf4155b3

</details>

<details>
<summary>Desktop</summary>

https://github.com/Expensify/App/assets/113963320/afe43929-8e08-4e96-bb87-2d67ce5d247a

</details>

<details>
<summary>iOS</summary>

https://github.com/Expensify/App/assets/113963320/580fea1c-365a-4062-9510-c2ae369eab54

</details>

<details>
<summary>Android</summary>

https://github.com/Expensify/App/assets/113963320/95fc8dc1-82f6-46c2-bd11-8426e5744a3d

</details>

### CODE REVIEW

File: src/libs/Middleware/SaveResponseInOnyx.js [hunk 1]
```diff
@@ -14,50 +14,52 @@ const requestsToIgnoreLastUpdateID = ['OpenApp', 'ReconnectApp', 'GetMissingOnyx
...
-        // we don't need to store anything here.
-        if (!onyxUpdates && !request.successData && !request.failureData) {
-            return Promise.resolve(response);
-        }
-
-        // If there is an OnyxUpdate for using memory only keys, enable them
-        _.find(onyxUpdates, ({key, value}) => {
-            if (key !== ONYXKEYS.IS_USING_MEMORY_ONLY_KEYS || !value) {
-                return false;
+    return requestResponse
```
Comment (abdulrahuman5196) [inline src/libs/Middleware/SaveResponseInOnyx.js:17]: @tienifr Why are having this much diff here? I know we are only adding catch to this? 

Can we remove the diff? Any specific reason to have?

File: src/libs/Middleware/SaveResponseInOnyx.js [hunk 2]
```diff
@@ -14,50 +14,52 @@ const requestsToIgnoreLastUpdateID = ['OpenApp', 'ReconnectApp', 'GetMissingOnyx
...
+            // Save the update IDs to Onyx so they can be used to fetch incremental updates if the client gets out of sync from the server
+            OnyxUpdates.saveUpdateInformation(responseToApply);
+
+            // Ensure the queue is paused while the client resolves the gap in onyx updates so that updates are guaranteed to happen in a specific order.
+            return Promise.resolve({
+                ...response,
+                shouldPauseQueue: true,
+            });
+        })
+        .catch(console.error);
```
Comment (abdulrahuman5196) [inline src/libs/Middleware/SaveResponseInOnyx.js:62]: Kindly add small meaningful tags with error, like `Got exception while saving response in onyx` @tienifr

**Comment by abdulrahuman5196** [comment]
Thank you checking

**Comment by abdulrahuman5196** [comment]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>






https://github.com/Expensify/App/assets/46707890/2efb89e7-29d8-49f2-a825-5d493acb4455


</details>

<details>
<summary>Mobile Web - Chrome</summary>




https://github.com/Expensify/App/assets/46707890/4daf5145-8290-4617-bb58-badeaa882a21


</details>

<details>
<summary>Mobile Web - Safari</summary>







https://github.com/Expensify/App/assets/46707890/9579a1da-4358-4616-812e-ca97d1338704


</details>

<details>
<summary>Desktop</summary>





https://github.com/Expensify/App/assets/46707890/89419139-cb65-4bdb-8cda-2be97c96c503



</details>

<details>
<summary>iOS</summary>






https://github.com/Expensify/App/assets/46707890/0f281d4d-13dc-49c8-a2fd-57cdb52775e7

</details>

<details>
<summary>Android</summary>






</details>

**Review by mountiny — APPROVED** [review]
Thanks for quick fix

File: src/libs/Middleware/SaveResponseInOnyx.js (hunk 1 above)
Comment (abdulrahuman5196) [reply src/libs/Middleware/SaveResponseInOnyx.js:17]: Thank you. I also verified. No changes.

**Review by abdulrahuman5196 — APPROVED** [review]
Changes looks good and works well. Reviewers checklist is also complete.

All yours. @mountiny 

🎀 👀 🎀
C+ Reviewed

**Review by mountiny — APPROVED** [review]
Thanks everyone!

File: src/libs/Middleware/SaveResponseInOnyx.js [hunk 3]
```diff
@@ -14,50 +14,54 @@ const requestsToIgnoreLastUpdateID = ['OpenApp', 'ReconnectApp', 'GetMissingOnyx
...
-            ...response,
-            shouldPauseQueue: true,
+            // Ensure the queue is paused while the client resolves the gap in onyx updates so that updates are guaranteed to happen in a specific order.
+            return Promise.resolve({
+                ...response,
+                shouldPauseQueue: true,
+            });
+        })
+        .catch((err) => {
+            console.error('Got exception while saving response in Onyx', err);
```
Comment (blazejkustra) [inline src/libs/Middleware/SaveResponseInOnyx.js:63]: This `catch` causes a regression, it suppresses network errors such as 'Failed to fetch', but I believe it shouldn't.

Unit tests defined in `APITest.js` are not failing due to a TypeError which is always raised in `SequentialQueue` file [here](https://github.com/Expensify/App/blob/5fcdfd9ebd187c11cab05c826ea10afec354b2fc/src/libs/Network/SequentialQueue.js#L74):

First an error is suppressed in `SaveResponseInOnyx`:
```ts
  console.error
    Got exception while saving response in Onyx Error: Failed to fetch
        at /tests/unit/APITest.js:90:40

      52 |         })
      53 |         .catch((err) => {
    > 54 |             console.error('Got exception while saving response in Onyx', err);
         |                     ^
      55 |         });
      56 |
      57 | export default SaveResponseInOnyx;
```

Later an error is raised as `shouldPauseQueue` isn't defined on `undefined` response:
```ts
    error1 =  TypeError: Cannot read properties of undefined (reading 'shouldPauseQueue')
        at /src/libs/Network/SequentialQueue.ts:73:26)

      at log (src/libs/Network/SequentialQueue.ts:81:21)
```

cc @mountiny @tienifr

Comment (mountiny) [reply src/libs/Middleware/SaveResponseInOnyx.js:63]: Thanks!

File: src/libs/Middleware/HandleUnusedOptimisticID.ts [hunk 4]
```diff
@@ -10,6 +10,10 @@ const handleUnusedOptimisticID: Middleware = (requestResponse, request, isFromSe
         const responseOnyxData = response?.onyxData ?? [];
         responseOnyxData.forEach((onyxData) => {
             const key = onyxData.key;
+            if (!key) {
+                return;
+            }
+
             if (!key.startsWith(ONYXKEYS.COLLECTION.REPORT)) {
```
Comment (blazejkustra) [inline src/libs/Middleware/HandleUnusedOptimisticID.ts:17]: FYI optional chaining is allowed in TS:
```suggestion
            if (!key?.startsWith(ONYXKEYS.COLLECTION.REPORT)) {
```

**Pull request merged**